src/sgidClient.ts

```diff
@@ -0,0 +1,233 @@
+
+  /**
+   * Initialises an SgidClient instance.
+   * @param params Constructor arguments
+   * @param params.clientId Client ID provided during client registration
+   * @param params.clientSecret Client secret provided during client registration
+   * @param params.privateKey Client private key provided during client registration
+   * @param params.redirectUri Redirection URI for user to return to your application
+   * after login. If not provided in the constructor, this must be provided to the
+   * authorizationUrl and callback functions.
+   * @param params.hostname Hostname of OpenID provider (sgID). Defaults to
+   * https://api.id.gov.sg.
+   */
+  constructor({
+    clientId,
+    clientSecret,
+    privateKey,
+    redirectUri,
+    hostname = 'https://api.id.gov.sg',
+  }: SgidClientParams) {
+    /**
+     * Note that issuer is appended with version number only from v2 onwards
+     */
+    const issuer = new URL(hostname).origin + `/v${API_VERSION}`
+
+    const { Client } = new Issuer({
+      issuer,
+      authorization_endpoint: `${issuer}/oauth/authorize`,
+      token_endpoint: `${issuer}/oauth/token`,
+      userinfo_endpoint: `${issuer}/oauth/userinfo`,
+      jwks_uri: `${issuer}/.well-known/jwks.json`,
+    })
+
+    this.sgID = new Client({
+      client_id: clientId,
+      client_secret: clientSecret,
+      redirect_uris: redirectUri ? [redirectUri] : undefined,
+      id_token_signed_response_alg: SGID_SIGNING_ALG,
+      response_types: SGID_SUPPORTED_FLOWS,
+      token_endpoint_auth_method: SGID_AUTH_METHOD,
+    })
+
+    /**
+     * For backward compatibility with pkcs1
+     */
+    if (privateKey.startsWith('-----BEGIN RSA PRIVATE KEY-----')) {
+      this.privateKey = convertPkcs1ToPkcs8(privateKey)
+    } else {
+      this.privateKey = privateKey
+    }
+
+    /**
+     * To rectify instances where an extra '\' escape character is added to the newline
+     */
+    this.privateKey = this.privateKey.replace(/\\n/gm, '\n')
+  }
+
+  /**
+   * Generates authorization url to redirect end-user to sgID login page.
+   * @param state A string which will be passed back to your application once the end-user
+   * logs in. You should use this to prevent cross-site request forgery attacks (see
+   * https://www.rfc-editor.org/rfc/rfc6749#section-10.12). You can also use this to
+   * track per-request state.
+   * @param scope Array or space-separated scopes. 'openid' must be provided as a scope.
+   * Defaults to 'myinfo.nric_number openid'.
+   * @param nonce Unique nonce for this request. If this param is undefined, a nonce is generated
+   * and returned. To prevent this behaviour, specify null for this param.
+   * @param redirectUri The redirect URI used in the authorization request. If this param is provided,
+   * it will be used over the redirect URI provided in the SgidClient constructor. The redirect URI
+   * must be provided in at least the SgidClient constructor or this function.
```

Comment:
hm think we need to rethink the English here. maybe "If not provided in the constructor, the redirect URI must be provided here."

Reply:
sounds good!